In [2]:
from writereadfile import *

In [5]:
import networkx as nx

In [9]:
creators=load_json('creators') #the data in 'creators' and 'distributors' contains all the users for all the 1036, NOTE not unique users
distributors=load_json('distributors')
tt_string=load_json('trending_topics_list368')

In [9]:
print(len(creators))
print(len(distributors))
print(len(tt_string))

56125
129134
368


In [6]:
creators[0]

[240811172,
 'CyrusCouturexx',
 '#mileyonsnl',
 'e86dba234c7429d9aea70c5f104dbebc',
 44252936221622272]

In [7]:
distributors[0]

[64572560,
 'ToxiicTara',
 '#mileyonsnl',
 'e86dba234c7429d9aea70c5f104dbebc',
 44252713759948801]

# RBC: Retweeting Graphs per trending topic

368 trending topics done for Retweeting Graph RBC method

In [10]:
for topic in tt_string:
    G=nx.DiGraph()
    
    cr=[(u[0],u[1],u[2],u[3],u[4]) for u in creators if u[2]==topic]
    di=[(u[0],u[1],u[2],u[3],u[4]) for u in distributors if u[2]==topic]
    
    cr_ids=[u[0] for u in cr]
    di_ids=[u[0] for u in di]
    repetidos=list(set(di_ids).intersection(set(cr_ids)))  #eliminate the distributors who were found as CREATORS
    
    di=[(u[0],u[1],u[2],u[3],u[4]) for u in di if u[0] not in repetidos]
    
    retweeting_pairs=[]
    for creator in cr:
        for distributor in di:
            if distributor[4] == creator[4]: #if the distributor retweeted the tweet t, then join the distributor with the creator
                retweeting_pairs.append((distributor[0], creator[0]))
    retweeting_pairs=list(set(retweeting_pairs))
    save_json('Retweeting_graphs368/'+topic, retweeting_pairs)
              
    nodes = list(set([n1 for n1, n2 in retweeting_pairs] + [n2 for n1, n2 in retweeting_pairs]))
    G.add_nodes_from(nodes)
    G.add_edges_from(retweeting_pairs)
    nameFile='Retweeting_graphs368/'+topic+'.gexf'
    nx.write_gexf(G,nameFile)
    G.clear()

# Following Graphs per trending topic for CREATORS

Create following graph among Creators

In [11]:
for topic in tt_string:
    G=nx.DiGraph()
    nameFile='FollowersCreators_perTrendingTopic/'+topic
    topic_folowers= load_json(nameFile)
    
    cr=[(u[0],u[1],u[2],u[3],u[4]) for u in creators if u[2]==topic]
    cr_ids=[u[0] for u in cr]
    
    following_pairs_user=[]
    following_pairs=[]
    
    for user in topic_folowers:
        following_pairs_user=[(inf, user['user_id']) for inf in cr_ids if inf in user['followers']]
        following_pairs+=following_pairs_user

    following_pairs=list(set(following_pairs))
    save_json('Following_graphs_Creators368/'+topic, following_pairs)
              
    nodes = list(set([n1 for n1, n2 in following_pairs] + [n2 for n1, n2 in following_pairs]))
    G.add_nodes_from(nodes)
    G.add_edges_from(following_pairs)
    nameFile='Following_graphs_Creators368/'+topic+'.gexf'
    nx.write_gexf(G,nameFile)
    G.clear()
                

# Following graphs for DISTRIBUTORS

Create following graph among Distributors

In [12]:
users_ids=load_json('unique_users_Creators_Distrib_perTopic368')

In [15]:
for topic in tt_string:
    G=nx.DiGraph()
    nameFile='FollowersDistributors_perTrendingTopic/'+topic
    topic_folowers= load_json(nameFile)
    
    di=[u['distributors'] for u in users_ids if u['topic']==topic][0]
    
    following_pairs_user=[]
    following_pairs=[]
    
    for user in topic_folowers:
        following_pairs_user=[(susc, user['user_id']) for susc in di if susc in user['followers']]
        following_pairs+=following_pairs_user

    following_pairs=list(set(following_pairs))
    save_json('Following_graphs_Distributors368/'+topic, following_pairs)
              
    nodes = list(set([n1 for n1, n2 in following_pairs] + [n2 for n1, n2 in following_pairs]))
    G.add_nodes_from(nodes)
    G.add_edges_from(following_pairs)
    nameFile='Following_graphs_Distributors368/'+topic+'.gexf'
    nx.write_gexf(G,nameFile)
    G.clear()

# TreToC graphs or OUR method graphs:   CreatorFollowing  U DistributorsFollowing  U  Distributors-Creators_Retweeting 

In [7]:
from writereadfile import *
import networkx as nx

In [16]:
for topic in tt_string:
    G=nx.DiGraph()
    Fcr=load_json('Following_graphs_Creators368/'+topic)
    Fcr=[(i[0], i[1]) for i in Fcr]
    Fco=load_json('Following_graphs_Distributors368/'+topic)
    Fco=[(i[0], i[1]) for i in Fco]
    Ret=load_json('Retweeting_graphs368/'+topic)
    Ret=[(i[0], i[1]) for i in Ret]
    
    print(len(Fcr+Fco+Ret))
    #print(Fcr+Fco+Ret)
    OUR=list(set(Fcr+Fco+Ret))
    print(len(OUR))
    
    save_json('Our_Method_TretocGraphs368/'+topic, OUR)
              
    nodes = list(set([n1 for n1, n2 in OUR] + [n2 for n1, n2 in OUR]))
    G.add_nodes_from(nodes)
    G.add_edges_from(OUR)
    nameFile='Our_Method_TretocGraphs368/'+topic+'.gexf'
    nx.write_gexf(G,nameFile)
    G.clear()

555
555
54
54
434
434
650
650
586
586
107
107
136
136
225
225
455
455
9
9
257
257
153
153
71
71
526
526
52
52
37
37
391
391
173
173
39
39
73
73
20
20
140
140
1
1
738
738
8
8
74
74
237
237
86
86
135
135
58
58
333
333
100
100
81
81
1441
1441
341
341
434
434
28
28
53
53
133
133
502
502
30
30
893
893
72
72
40
40
320
320
6
6
203
203
146
146
36
36
50
50
211
211
200
200
261
261
172
172
2286
2286
68
68
35
35
1597
1597
440
440
63
63
36
36
137
137
431
431
188
188
9
9
127
127
1584
1584
382
382
695
695
295
295
223
223
7
7
5483
5483
97
97
164
164
436
436
697
697
49
49
58
58
2348
2348
202
202
20
20
91
91
144
144
110
110
8
8
138
138
78
78
409
409
107
107
1958
1958
117
117
35
35
250
250
14
14
125
125
119
119
89
89
0
0
44
44
3
3
134
134
82
82
16
16
43
43
200
200
1260
1260
117
117
144
144
232
232
81
81
43
43
365
365
25
25
128
128
68
68
186
186
411
411
195
195
135
135
108
108
23
23
53
53
21
21
27
27
15
15
199
199
122
122
252
252
2999
2999
76
76
677
677
46
46
166
166
136
136
350
350
979
979
365
365
481
48